pkg/hubble/container/ring.go

```diff
@@ -31,6 +32,67 @@ var (
 	ErrInvalidRead = errors.New("read position is no longer valid")
 )
 
+// capacity implements Capacity.
+type capacity uint16
+
+// Cap returns the actual capacity.
+func (c capacity) Cap() capacity {
+	return c
+}
+
+// AsInt returns the actual capacity as an int.
+func (c capacity) AsInt() int {
+	return int(c)
+}
+
+// Capacity is the interface that wraps Cap.
+type Capacity interface {
+	// Cap returns the actual capacity.
+	Cap() capacity
+	// AsInt returns the actual capacity as an int.
+	AsInt() int
+}
+
+// CapacityN represent possible buffer capacities for Ring where N is the
+// actual capacity.
+const (
+	Capacity1 capacity = 1<<(iota+1) - 1
+	Capacity3
+	Capacity7
+	Capacity15
+	Capacity31
+	Capacity63
+	Capacity127
+	Capacity255
+	Capacity511
+	Capacity1023
+	Capacity2047
+	Capacity4095
+	Capacity8191
+	Capacity16383
+	Capacity32767
+	Capacity65535
+)
+
+// NewCapacity creates a new Capacity from n.
+// The value of n MUST satisfy n=2^i -1 for i = [1, 16]; ie:
+//
+//     1, 3, 7, ..., 2047, 4095, ..., 65535
+//
+// Constants CapacityN represent all possible values of n and are valid
+// Capacity that can be provided to NewRing.
+func NewCapacity(n int) (Capacity, error) {
+	switch {
+	case n > int(^capacity(0)):
+		return nil, fmt.Errorf("invalid capacity: too large: %d", n)
+	case n > 0:
+		if n&(n+1) == 0 {
+			return capacity(n), nil
+		}
+	}
+	return nil, fmt.Errorf("invalid capacity: %d", n)
+}
+
```

Comment:
Is a new `capacity` type really needed? How about using an `int` and having a `validCapacity` function:

```go
func validCapacity(capacity int) bool {
    return capacity > 0 && capacity == NextPowerOfTwo(capacity) - 1
}

...

    if !validCapacity(capacity) {
        return fmt.Errorf("invalid capacity, not a power of two minus one: %d", capacity)
    }
```
```

Reply:
The whole idea is that with the new type, it's impossible to specify a value outside the range of what is specified which makes it impossible to misuse (note that the current type is also bounded to `65535`). If we revert to using an `int` for `NewRing`, we should then also add some checks on the provided value and modify the constructor's signature to return an error. (eg: a negative int value could cause a panic because a type cast from negative `int` to `uint64` may result in a value too large for `make` which means `panic: runtime error: makeslice: len out of range`)

Reply:
Using `uint16` for capacity still allows many unrepresentable states, so I don't think there is a significant gain in safety in using `uint16` over `int`, but feel free to ignore my personal preferences :)

Reply:
The safety of the new type is not in that it's a `uint16` but rather that it cannot be misused outside the `container` package. Indeed, the `container.Capacity` interface cannot be implemented outside the `container` package because the `Cap() capacity` method returns an internal type. The only option to use/create a new `Capacity` when outside the `container` package is to use one of the `container.CapacityN` constant or the `container.NewCapacity` constructor which enforces a valid capacity.

Now, yes, I could define `capacity` as `type capacity int` but that would make very little difference in practice as one would still need a type cast to "convert" back to an `int` for `AsInt()` or within the `NewRing` constructor.

Reply:
Ah, I see. Thank you for the explanation.